### PR TITLE
dask: `Data` collapse methods and functions

### PR DESCRIPTION
**Pull request by davidhassell**
Another hefty PR! Hopefully self explanatory, but these notes might be useful:

* Added `Data.square` and `Data.sqrt` because they were useful.
* Made some changes to `Data.stats` and `Data.mean_of_upper_decile` before realising that they can't be tested yet. So these will be dealt with in separate PRs. Feel free to ignore these methods.
* Needs dask v2022.03.0 which should be out later today :)
* Tidied up (i.e. removed) some redundant MPI stuff.

### CODE REVIEW

**Comment by davidhassell** [comment]
Hang on - forgot a test ....

**Comment by davidhassell** [comment]
OK - good to go.

**Comment by davidhassell** [comment]
Another note:

* I got rid of `Data._map_blocks` because a) it wasn't used very often, and b) to use it you really have to check what it does internally, and since it's only ~3 lines it made sense to me to in-line it.

File: cf/docstring/docstring.py [hunk 1]
```diff
@@ -245,6 +239,108 @@
                   domain axis. If the vertical axis does not appear in
                   the computed non-parametric coodinates then this an
                   empty tuple.""",
+    # collapse axes
+    "{{collapse axes: (sequence of) `int`, optional}}": """axes: (sequence of) `int`, optional
+                The axes to be collapsed. By default all axes are
+                collapsed, resulting in output with size 1. Each axis
+                is identified by its integer position. If *axes* is an
+                empty sequence then the collapse is applied to each
+                scalar element and the reuslt has the same shape as
```
Comment (sadielbartholomew) [inline cf/docstring/docstring.py:248]:
```suggestion
                scalar element and the result has the same shape as
```

File: docs/source/field_analysis.rst [hunk 2]
```diff
@@ -131,84 +131,116 @@ the new cell method construct (if one is created).
...
+                              .. math::  m_n=\min\{x_0, \ldots, x_N\}
                                    
-``'maximum_absolute_value'``  The maximum of the absolute values.       ``maximum_absolute_value``
+``'maximum_absolute_value'``  The maximum of the non-missing absolute   ``maximum_absolute_value``
+                              values.
+
+                              .. math:: \max\{|x_0|, \ldots, |x_N|\}
                           
-``'minimum_absolute_value'``  The minimum of the absolute values.       ``minimum_absolute_value``
+``'minimum_absolute_value'``  The minimum of the absolute absolute      ``minimum_absolute_value``
```
Comment (sadielbartholomew) [inline docs/source/field_analysis.rst:150]:
```suggestion
``'minimum_absolute_value'``  The minimum of the non-missing absolute   ``minimum_absolute_value``
```

File: cf/data/data.py [hunk 3]
```diff
@@ -8028,143 +7329,76 @@ def mean(
...
         :Parameters:
 
-            axes : (sequence of) int, optional
+            {{collapse axes: (sequence of) `int`, optional}}
+
+            {{weights: data_like, `dict`, or `None`, optional}}
+
+            {{collapse squeeze: `bool`, optional}}
+
+            {{mtol: number, optional}
```
Comment (sadielbartholomew) [inline cf/data/data.py:7397]:
```suggestion
            {{mtol: number, optional}}
```

There's about twenty cases of this throughout `data.data` here, as I see from a `grep`. I won't pose them all as review 'suggestions'! I suggest addressing them in batch via `sed` or similar...

**Review by sadielbartholomew — COMMENTED** [review]
I'm mid-way through reviewing and everything seems good so far. Thought I'd raise some typos in the meantime, and if you can address the conflict resolutions that would help me to ensure I am reviewing the up-to-date code. Thanks.

**Review by sadielbartholomew — COMMENTED** [review]
Overall this is fantastic. The updated code is well-structured and even copes sensibly across all of the statistics (bar those two methods you say are out-of-scope for this PR) for the slightly evil case of a fully-masked array. Great commitment to thorough documentation, too.

I've posed a few questions on minor aspects and suggested fixes for inevitable typos, one of which is in the functional code so is quite crucial to update. Thanks.

File: docs/source/field_analysis.rst [hunk 4]
```diff
@@ -131,84 +131,116 @@ the new cell method construct (if one is created).
 ============================  ========================================  ==========================
 Method                        Description                               Cell method
 ============================  ========================================  ==========================
-``'maximum'``                 The maximum of the values.                ``maximum``
+``'sample_size'``             The sample size, :math:`N`, equal to the  ``point``
+			      number of non-missing values.
+                              
+``'maximum'``                 The maximum of the non-missing values.    ``maximum``
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply docs/source/field_analysis.rst:137]: Took me a moment to work out what's going on here. The "maximum" here is not a Field construct method, rather a CF cell method 'method'. This is a case of the docs being not wrong, but a bit obfuscated! I'll see if I can re-word things ...

File: cf/data/data.py [hunk 5]
```diff
@@ -13488,84 +12943,241 @@ def section(
...
+
+        """
+        d = _inplace_enabled_define_and_cleanup(self)
+        dx = d.to_dask_array()
+        dx = da.sqrt(dx, dtype=dtype)
+        d._set_dask(dx, reset_mask_hardness=False)
+
+        units = d.Units
+        if units:
+            d.override_units(units ** 0.5, inplace=True)
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/data/data.py:13060]: I definitely think that being clear on units is the best way - I'll put in a better message ...

File: cf/test/test_Data.py [hunk 6]
```diff
@@ -3919,6 +3129,679 @@ def test_Data_rechunk(self):
...
+
+        for x, r in zip((d, e), ("i8", "f8")):
+            for func in (
+                x.integral,
+                x.sum,
+                x.sum_of_squares,
+            ):
+                self.assertEqual(func().dtype, r)
+
+        for x, r in zip((d, e), ("f8", "f8")):
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/test/test_Data.py:3777]: It does look odd, but it's testing that both integer and float data collapse to a float result. I'll put in  a comment:  https://github.com/NCAS-CMS/cf-python/pull/356/commits/441f74bb6585e96c6aef5dd47cd615a8d623ba1e

File: cf/data/data.py [hunk 7]
```diff
@@ -5076,913 +5050,163 @@ def __pos__(self):
...
+    def _cyclic(self):
+        self._custom["_cyclic"] = _empty_set
 
-        :Returns:
+    @property
+    def _HDF_chunks(self):
+        """The HDF chunksizes.
 
-            `Data` or `None`
+        DO NOT CHANGE IN PLACE.
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/data/data.py]: Oops - the `_HDF_chunks` property should have moved to `mixin/deprecated.py` along with `HDF_chunks` in #338 : https://github.com/NCAS-CMS/cf-python/pull/356/commits/0d87f44f7b5c96bae0493b602eee7a660403354b

File: cf/data/data.py [hunk 8]
```diff
@@ -5076,913 +5050,163 @@ def __pos__(self):
...
-                result as axes with size 1. In this case the result will
-                broadcast correctly against the original array. By default
-                collapsed axes are removed.
+        Contains a `set` that identifies which axes are cyclic (and
+        therefore allow cyclic slicing). The set contains a subset of
+        the axis identifiers defined by the `_axes` attribute.
 
-            weights: *optional*
+        .. warning:: Never change the value of the `_cyclic` attribute
+                     in-place.
```
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply cf/data/data.py:5096]: Thanks for clarifying, David. That makes sense and in light of it a warning seems best. Marking as resolved.

File: cf/test/test_Data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply cf/test/test_Data.py:3777]: Ah, thanks for clarifying within the tests themselves. That will be useful for anyone like myself who sees the `("iX", "fX")` pattern from other tests and wonders why this one is different!

File: cf/data/data.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/data/data.py]: https://github.com/NCAS-CMS/cf-python/pull/356/commits/82d3508e7872365f531cabd469a550b66c2ebd91

File: docs/source/field_analysis.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidhassell) [reply docs/source/field_analysis.rst:138]: How about https://github.com/NCAS-CMS/cf-python/pull/356/commits/60fcb17b57043379eb0c2763b88bda1f97a01e62 ?

[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply docs/source/field_analysis.rst:138]: That makes it nice and clear. Thanks. Marking as resolved. Let's merge!

**Pull request merged**